File: scripts/extract-theme.ts
#!/usr/bin/env bun
/**
 * primaviz-theme-cli — Extract design tokens from any CSS file and generate
 * primaviz-compatible theme files (Typst and/or JSON).
 *
 * Usage:
 *   bun run scripts/extract-theme.ts <css-file> [options]
 *
 * Options:
 *   --out-dir <dir>       Output directory (default: ./typst)
 *   --format <fmt>        Output format: typst, json, both (default: both)
 *   --name <name>         Theme name prefix (default: theme)
 *   --dark-selector <s>   CSS selector for dark mode (default: .dark)
 *   --help                Print usage information
 */

import { readFileSync, writeFileSync, mkdirSync, existsSync } from "fs";
import { resolve, basename } from "path";
import { parse, formatHex, converter } from "culori";

// ─── Argument parsing ────────────────────────────────────────────────────────

interface CliOptions {
  cssFile: string;
  outDir: string;
  format: "typst" | "json" | "both";
  name: string;
  darkSelector: string;
}

function printHelp(): void {
  console.log(`
primaviz-theme-cli — Extract design tokens from CSS and generate primaviz theme files.

Usage:
  bun run scripts/extract-theme.ts <css-file> [options]

Arguments:
  <css-file>            Path to a CSS file containing design tokens (oklch, hsl, hex)

Options:
  --out-dir <dir>       Output directory (default: ./typst)
  --format <fmt>        Output format: typst, json, both (default: both)
  --name <name>         Theme name prefix (default: theme)
  --dark-selector <s>   CSS selector for dark mode (default: .dark)
  --help                Print this help message

Examples:
  bun run scripts/extract-theme.ts src/index.css
  bun run scripts/extract-theme.ts styles/globals.css --name shadcn --format typst
  bun run scripts/extract-theme.ts theme.css --out-dir output --dark-selector '[data-theme="dark"]'
`);
}

function parseArgs(argv: string[]): CliOptions | null {
  const args = argv.slice(2); // skip bun and script path

  if (args.length === 0 || args.includes("--help")) {
    printHelp();
    return null;
  }

  // First positional argument is the CSS file
  let cssFile = "";
  let outDir = "./typst";
  let format: "typst" | "json" | "both" = "both";
  let name = "theme";
  let darkSelector = ".dark";

  let i = 0;
  while (i < args.length) {
    const arg = args[i];
    if (arg === "--out-dir") {
      outDir = args[++i] || outDir;
    } else if (arg === "--format") {
      const fmt = args[++i];
      if (fmt === "typst" || fmt === "json" || fmt === "both") {
        format = fmt;
      } else {
        console.error(`Error: Invalid format "${fmt}". Must be typst, json, or both.`);
        process.exit(1);
      }
    } else if (arg === "--name") {
      name = args[++i] || name;
    } else if (arg === "--dark-selector") {
      darkSelector = args[++i] || darkSelector;
    } else if (!arg.startsWith("--")) {
      cssFile = arg;
    } else {
      console.error(`Error: Unknown option "${arg}". Use --help for usage.`);
      process.exit(1);
    }
    i++;
  }

  if (!cssFile) {
    console.error("Error: No CSS file specified. Use --help for usage.");
    process.exit(1);
  }

  return { cssFile: resolve(cssFile), outDir: resolve(outDir), format, name, darkSelector };
}

// ─── Color conversion ────────────────────────────────────────────────────────

const toRgb = converter("rgb");

/**
 * Convert any CSS color (oklch, hsl, rgb, hex) to a hex string.
 * Handles alpha by blending onto a background color.
 */
function colorToHex(raw: string, bgHex = "#FFFFFF"): string | null {
  const color = parse(raw);
  if (!color) return null;

  if (color.alpha != null && color.alpha < 1) {
    const fg = toRgb(color);
    const bg = toRgb(parse(bgHex)!);
    if (!fg || !bg) return formatHex(color);
    const a = color.alpha;
    return formatHex({
      mode: "rgb" as const,
      r: fg.r * a + bg.r * (1 - a),
      g: fg.g * a + bg.g * (1 - a),
      b: fg.b * a + bg.b * (1 - a),
    });
  }

  return formatHex(color);
}

// ─── CSS parsing ─────────────────────────────────────────────────────────────

/**
 * Extract CSS custom properties from a CSS block string.
 */
function extractVars(block: string): Record<string, string> {
  const vars: Record<string, string> = {};
  const re = /--([\w-]+):\s*(.+?);/g;
  let m;
  while ((m = re.exec(block)) !== null) {
    vars[m[1]] = m[2].trim();
  }
  return vars;
}

/**
 * Extract the contents of the :root { ... } block.
 */
function extractRootBlock(css: string): string {
  const match = css.match(/:root\s*\{([^}]+)\}/);
  return match ? match[1] : "";
}

/**
 * Extract the contents of a dark mode block by selector.
 * Handles class selectors (.dark), attribute selectors ([data-theme="dark"]),
 * and media queries (@media (prefers-color-scheme: dark)).
 */
function extractDarkBlock(css: string, selector: string): string {
  // Try the selector directly (e.g., .dark { ... })
  const escaped = selector.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const directMatch = css.match(new RegExp(escaped + "\\s*\\{([^}]+)\\}"));
  if (directMatch) return directMatch[1];

  // Try as media query for prefers-color-scheme
  if (selector === "@media (prefers-color-scheme: dark)") {
    const mediaMatch = css.match(
      /@media\s*\(\s*prefers-color-scheme\s*:\s*dark\s*\)\s*\{[^}]*:root\s*\{([^}]+)\}/
    );
    if (mediaMatch) return mediaMatch[1];
    // Also try without :root nested inside
    const mediaMatch2 = css.match(
      /@media\s*\(\s*prefers-color-scheme\s*:\s*dark\s*\)\s*\{([^}]+)\}/
    );
    if (mediaMatch2) return mediaMatch2[1];
  }

  return "";
}

// ─── Token mapping ───────────────────────────────────────────────────────────

/**
 * Map of CSS custom property names to primaviz semantic token names.
 * The extractor looks for these properties in :root and dark blocks.
 */
const SEMANTIC_MAP: Record<string, string> = {
  foreground: "fg",
  "muted-foreground": "muted-fg",
  "primary-foreground": "primary-fg",
  background: "bg",
  border: "border-color",
  card: "card-bg",
  "card-foreground": "card-fg",
  primary: "primary",
  "primary-foreground": "primary-fg",
  secondary: "secondary",
  "secondary-foreground": "secondary-fg",
  muted: "muted-bg",
  accent: "accent",
  "accent-foreground": "accent-fg",
  destructive: "destructive",
  input: "input-border",
  ring: "ring",
};

/**
 * Detect chart color custom properties (--chart-1, --chart-2, etc.).
 * Returns sorted array of [index, cssPropertyName].
 */
function findChartProperties(vars: Record<string, string>): Array<[number, string]> {
  const charts: Array<[number, string]> = [];
  for (const key of Object.keys(vars)) {
    const m = key.match(/^chart-(\d+)$/);
    if (m) {
      charts.push([parseInt(m[1], 10), key]);
    }
  }
  return charts.sort((a, b) => a[0] - b[0]);
}

// ─── Theme extraction ────────────────────────────────────────────────────────

interface ThemeTokens {
  palette: string[];
  "text-color": string;
  "text-color-light": string;
  "text-color-inverse": string;
  background: string | null;
  "border-color": string;
  "border-radius": number;
  /** All extracted semantic colors for reference */
  all: Record<string, string>;
}

function extractThemeTokens(
  vars: Record<string, string>,
  bgHex: string,
  isDark: boolean
): ThemeTokens {
  // Convert all semantic tokens
  const converted: Record<string, string> = {};
  for (const [cssProp, tokenName] of Object.entries(SEMANTIC_MAP)) {
    const raw = vars[cssProp];
    if (!raw) continue;
    const hex = colorToHex(raw, bgHex);
    if (hex) converted[tokenName] = hex.toUpperCase();
  }

  // Convert chart colors
  const chartProps = findChartProperties(vars);
  const palette: string[] = [];
  for (const [, cssProp] of chartProps) {
    const raw = vars[cssProp];
    if (!raw) continue;
    const hex = colorToHex(raw, bgHex);
    if (hex) palette.push(hex.toUpperCase());
  }

  // Extract radius
  const radiusRaw = vars["radius"];
  let radiusPt = 8; // default
  if (radiusRaw) {
    const rem = parseFloat(radiusRaw);
    if (!isNaN(rem)) {
      radiusPt = Math.round(rem * 12); // approximate rem to pt
    }
  }

  return {
    palette,
    "text-color": converted["fg"] || (isDark ? "#FAFAFA" : "#0A0A0A"),
    "text-color-light": converted["muted-fg"] || (isDark ? "#8888AA" : "#888888"),
    "text-color-inverse": converted["primary-fg"] || "#FAFAFA",
    background: isDark ? (converted["bg"] || "#0A0A0A") : null,
    "border-color": converted["border-color"] || (isDark ? "#2A2A3E" : "#E2E2EE"),
    "border-radius": radiusPt,
    all: converted,
  };
}

// ─── Output generation ───────────────────────────────────────────────────────

function generateTypst(
  light: ThemeTokens,
  dark: ThemeTokens,
  themeName: string
): string {
  const lines: string[] = [
    `// Auto-generated primaviz theme — do not edit by hand`,
    `// Generated by primaviz-theme-cli`,
    "",
  ];

  // Light theme
  const varName = themeName.replace(/[^a-zA-Z0-9-]/g, "-");
  lines.push(`#let ${varName} = (`);
  lines.push(`  palette: (`);
  for (let i = 0; i < light.palette.length; i++) {
    lines.push(`    rgb("${light.palette[i]}"),  // --chart-${i + 1}`);
  }
  lines.push(`  ),`);
  lines.push(`  text-color: rgb("${light["text-color"]}"),`);
  lines.push(`  text-color-light: rgb("${light["text-color-light"]}"),`);
  lines.push(`  text-color-inverse: rgb("${light["text-color-inverse"]}"),`);
  lines.push(`  background: none,`);
  lines.push(`  border: 0.75pt + rgb("${light["border-color"]}"),`);
  lines.push(`  axis-stroke: 0.5pt + rgb("${light["border-color"]}"),`);
  lines.push(`  grid-stroke: 0.3pt + rgb("${light["border-color"]}"),`);
  lines.push(`  show-grid: true,`);
  lines.push(`  border-radius: ${light["border-radius"]}pt,`);
  lines.push(`)`);
  lines.push("");

  // Dark theme
  lines.push(`// Dark mode variant`);
  lines.push(`#let ${varName}-dark = (`);
  lines.push(`  palette: (`);
  for (let i = 0; i < dark.palette.length; i++) {
    lines.push(`    rgb("${dark.palette[i]}"),  // --chart-${i + 1}`);
  }
  lines.push(`  ),`);
  lines.push(`  text-color: rgb("${dark["text-color"]}"),`);
  lines.push(`  text-color-light: rgb("${dark["text-color-light"]}"),`);
  lines.push(`  text-color-inverse: rgb("${dark["text-color-inverse"]}"),`);
  lines.push(`  background: rgb("${dark.background || "#0A0A0A"}"),`);
  lines.push(`  border: 0.75pt + rgb("${dark["border-color"]}"),`);
  lines.push(`  axis-stroke: 0.5pt + rgb("${dark["border-color"]}"),`);
  lines.push(`  grid-stroke: 0.3pt + rgb("${dark["border-color"]}"),`);
  lines.push(`  show-grid: true,`);
  lines.push(`  border-radius: ${dark["border-radius"]}pt,`);
  lines.push(`)`);
  lines.push("");

  return lines.join("\n");
}

function generateJson(light: ThemeTokens, dark: ThemeTokens): string {
  const output = {
    light: {
      palette: light.palette,
      "text-color": light["text-color"],
      "text-color-light": light["text-color-light"],
      "text-color-inverse": light["text-color-inverse"],
      background: light.background,
      "border-color": light["border-color"],
      "border-radius": light["border-radius"],
    },
    dark: {
      palette: dark.palette,
      "text-color": dark["text-color"],
      "text-color-light": dark["text-color-light"],
      "text-color-inverse": dark["text-color-inverse"],
      background: dark.background,
      "border-color": dark["border-color"],
      "border-radius": dark["border-radius"],
    },
  };
  return JSON.stringify(output, null, 2) + "\n";
}

// ─── Main ────────────────────────────────────────────────────────────────────

function main(): void {
  const opts = parseArgs(process.argv);
  if (!opts) process.exit(0);

  // Read CSS file
  if (!existsSync(opts.cssFile)) {
    console.error(`Error: CSS file not found: ${opts.cssFile}`);
    process.exit(1);
  }
  const css = readFileSync(opts.cssFile, "utf-8");
  console.log(`Reading tokens from ${opts.cssFile}`);

  // Extract light mode variables from :root
  const rootBlock = extractRootBlock(css);
  const rootVars = extractVars(rootBlock);
  const rootVarCount = Object.keys(rootVars).length;

  if (rootVarCount === 0) {
    console.warn("Warning: No custom properties found in :root block.");
  } else {
    console.log(`Found ${rootVarCount} custom properties in :root`);
  }

  // Determine light background for alpha blending
  const lightBgRaw = rootVars["background"] || "#FFFFFF";
  const lightBgHex = colorToHex(lightBgRaw) || "#FFFFFF";

  // Extract light mode tokens
  const lightTokens = extractThemeTokens(rootVars, lightBgHex, false);

  // Extract dark mode variables
  const darkBlock = extractDarkBlock(css, opts.darkSelector);
  const darkVars = extractVars(darkBlock);
  const darkVarCount = Object.keys(darkVars).length;

  if (darkVarCount === 0) {
    console.warn(
      `Warning: No custom properties found for dark selector "${opts.darkSelector}".`
    );
    console.warn("Dark theme will use fallback values.");
  } else {
    console.log(
      `Found ${darkVarCount} custom properties in "${opts.darkSelector}"`
    );
  }

  // Determine dark background for alpha blending
  const darkBgRaw = darkVars["background"] || "oklch(0.145 0 0)";
  const darkBgHex = colorToHex(darkBgRaw) || "#0A0A0A";

  // Extract dark mode tokens
  const darkTokens = extractThemeTokens(darkVars, darkBgHex, true);

  // Ensure output directory exists
  if (!existsSync(opts.outDir)) {
    mkdirSync(opts.outDir, { recursive: true });
    console.log(`Created output directory: ${opts.outDir}`);
  }

  // Write output files
  if (opts.format === "typst" || opts.format === "both") {
    const typstPath = resolve(opts.outDir, `${opts.name}.typ`);
    const typstContent = generateTypst(lightTokens, darkTokens, opts.name);
    writeFileSync(typstPath, typstContent);
    console.log(
      `Wrote Typst theme (${lightTokens.palette.length} light + ${darkTokens.palette.length} dark palette colors) to ${typstPath}`
    );
  }

  if (opts.format === "json" || opts.format === "both") {
    const jsonPath = resolve(opts.outDir, `${opts.name}.json`);
    const jsonContent = generateJson(lightTokens, darkTokens);
    writeFileSync(jsonPath, jsonContent);
    console.log(`Wrote JSON tokens (light + dark) to ${jsonPath}`);
  }

  console.log("Done.");
}

main();

File: scripts/extract-theme.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.10"
# dependencies = ["coloraide"]
# ///
"""Extract design tokens from CSS and generate primaviz theme files.

Reads CSS custom properties (oklch, hsl, rgb, hex) from :root and .dark blocks,
converts to hex, and outputs primaviz-compatible theme files (.typ and/or .json).

Usage:
    uv run scripts/extract-theme.py <css-file> [options]

Options:
    --out-dir DIR         Output directory (default: ./typst)
    --format FORMAT       Output format: typst, json, both (default: both)
    --name NAME           Theme name prefix (default: theme)
    --dark-selector SEL   CSS selector for dark mode (default: .dark)

Examples:
    uv run scripts/extract-theme.py src/index.css
    uv run scripts/extract-theme.py styles.css --out-dir out --format json
    uv run scripts/extract-theme.py globals.css --name shadcn --dark-selector '[data-theme="dark"]'

Token mapping:
    --chart-1..N          → palette (array)
    --foreground          → text-color
    --muted-foreground    → text-color-light
    --primary-foreground  → text-color-inverse
    --background          → background
    --border              → border-color
    --radius              → border-radius
"""

import argparse
import json
import colorsys
import math
import os
import re
import sys


# ── Color conversion ──────────────────────────────────────────────────────────

def parse_hex(s):
    """Parse hex color to (r, g, b) floats 0-1."""
    s = s.strip().lstrip("#")
    if len(s) == 3:
        s = s[0]*2 + s[1]*2 + s[2]*2
    if len(s) == 6:
        return (int(s[0:2], 16) / 255, int(s[2:4], 16) / 255, int(s[4:6], 16) / 255, 1.0)
    if len(s) == 8:
        return (int(s[0:2], 16) / 255, int(s[2:4], 16) / 255, int(s[4:6], 16) / 255, int(s[6:8], 16) / 255)
    return None


def to_hex(r, g, b):
    """Convert (r, g, b) floats 0-1 to hex string."""
    return "#{:02X}{:02X}{:02X}".format(
        max(0, min(255, round(r * 255))),
        max(0, min(255, round(g * 255))),
        max(0, min(255, round(b * 255))),
    )


def parse_hsl(s):
    """Parse hsl/hsla(...) to (r, g, b, a) floats."""
    m = re.match(r"hsla?\(\s*([\d.]+)\s*[\s,]\s*([\d.]+)%?\s*[\s,]\s*([\d.]+)%?(?:\s*[/,]\s*([\d.]+%?))?\s*\)", s)
    if not m:
        return None
    h = float(m.group(1)) / 360
    s_val = float(m.group(2)) / 100
    l = float(m.group(3)) / 100
    a = 1.0
    if m.group(4):
        a_str = m.group(4).strip()
        a = float(a_str.rstrip("%")) / 100 if "%" in a_str else float(a_str)
    r, g, b = colorsys.hls_to_rgb(h, l, s_val)
    return (r, g, b, a)


def parse_oklch(s):
    """Parse oklch(...) to (r, g, b, a) floats via coloraide."""
    from coloraide import Color

    m = re.match(r"oklch\(\s*([\d.]+)\s+([\d.]+)\s+([\d.]+)(?:\s*/\s*([\d.]+%?))?\s*\)", s)
    if not m:
        return None

    L = float(m.group(1))
    C = float(m.group(2))
    h = float(m.group(3))
    a = 1.0
    if m.group(4):
        a_str = m.group(4).strip()
        a = float(a_str.rstrip("%")) / 100 if "%" in a_str else float(a_str)

    c = Color("oklch", [L, C, h])
    rgb = c.convert("srgb")
    return (max(0, min(1, rgb["red"])), max(0, min(1, rgb["green"])), max(0, min(1, rgb["blue"])), a)


def color_to_hex(raw, bg_hex="#FFFFFF"):
    """Convert any CSS color value to hex, blending alpha onto background."""
    raw = raw.strip()
    rgba = None

    if raw.startswith("oklch("):
        rgba = parse_oklch(raw)
    elif raw.startswith("hsl"):
        rgba = parse_hsl(raw)
    elif raw.startswith("#"):
        rgba = parse_hex(raw)
    elif raw.startswith("rgb"):
        m = re.match(r"rgba?\(\s*([\d.]+)\s*[\s,]\s*([\d.]+)\s*[\s,]\s*([\d.]+)(?:\s*[/,]\s*([\d.]+%?))?\s*\)", raw)
        if m:
            r, g, b = float(m.group(1)) / 255, float(m.group(2)) / 255, float(m.group(3)) / 255
            a = 1.0
            if m.group(4):
                a_str = m.group(4).strip()
                a = float(a_str.rstrip("%")) / 100 if "%" in a_str else float(a_str)
            rgba = (r, g, b, a)

    if rgba is None:
        return None

    r, g, b, a = rgba

    # Alpha blend onto background
    if a < 1.0:
        bg = parse_hex(bg_hex)
        if bg:
            r = r * a + bg[0] * (1 - a)
            g = g * a + bg[1] * (1 - a)
            b = b * a + bg[2] * (1 - a)

    return to_hex(r, g, b)


# ── CSS parsing ───────────────────────────────────────────────────────────────

def extract_vars(css, selector):
    """Extract CSS custom properties from a selector block."""
    # Handle both `:root` and arbitrary selectors like `.dark`
    escaped = re.escape(selector)
    pattern = re.compile(escaped + r"\s*\{([^}]+)\}", re.DOTALL)
    m = pattern.search(css)
    if not m:
        return {}

    block = m.group(1)
    props = {}
    for line in block.split(";"):
        line = line.strip()
        match = re.match(r"--([\w-]+)\s*:\s*(.+)", line)
        if match:
            props[match.group(1)] = match.group(2).strip()
    return props


def build_tokens(props, bg_hex="#FFFFFF"):
    """Convert CSS props dict to token dict."""
    tokens = {}

    # Palette: --chart-1, --chart-2, ...
    palette = []
    i = 1
    while f"chart-{i}" in props:
        hex_val = color_to_hex(props[f"chart-{i}"], bg_hex)
        if hex_val:
            palette.append(hex_val)
        i += 1
    if palette:
        tokens["palette"] = palette

    # Semantic tokens
    mapping = {
        "foreground": "text-color",
        "muted-foreground": "text-color-light",
        "primary-foreground": "text-color-inverse",
        "background": "background",
        "border": "border-color",
    }
    for css_key, token_key in mapping.items():
        if css_key in props:
            hex_val = color_to_hex(props[css_key], bg_hex)
            if hex_val:
                tokens[token_key] = hex_val

    # Background: null if white/transparent
    if "background" in tokens:
        bg = tokens["background"]
        if bg and bg.upper() in ("#FFFFFF", "#FFF"):
            tokens["background"] = None

    # Radius — convert to pt
    if "radius" in props:
        val = props["radius"].strip()
        m = re.match(r"([\d.]+)\s*(rem|px|pt)?", val)
        if m:
            num = float(m.group(1))
            unit = m.group(2) or "px"
            if unit == "rem":
                num = num * 16 * 0.75  # rem → px → pt
            elif unit == "px":
                num = num * 0.75  # px → pt
            # pt stays as-is
            tokens["border-radius"] = round(num, 1)

    return tokens


# ── Output generators ─────────────────────────────────────────────────────────

def generate_typst(name, light, dark):
    """Generate .typ file content with primaviz theme dicts."""
    lines = [
        f"// {name}.typ — auto-generated primaviz theme from CSS tokens",
        f"// Run: python scripts/extract-theme.py <css-file>",
        "",
    ]

    for label, tokens in (("", light), ("-dark", dark)):
        dict_name = f"{name.replace('-', '_')}" + ("_dark" if label else "")
        lines.append(f"#let {dict_name} = (")

        if "palette" in tokens:
            lines.append(f"  palette: (")
            for hex_val in tokens["palette"]:
                lines.append(f'    rgb("{hex_val}"),')
            lines.append(f"  ),")

        key_map = [
            ("text-color", "text-color"),
            ("text-color-light", "text-color-light"),
            ("text-color-inverse", "text-color-inverse"),
        ]
        for token_key, theme_key in key_map:
            if token_key in tokens:
                lines.append(f'  {theme_key}: rgb("{tokens[token_key]}"),')

        if "background" in tokens:
            bg = tokens["background"]
            if bg is None:
                lines.append("  background: none,")
            else:
                lines.append('  background: rgb("{}"),'.format(bg))

        if "border-color" in tokens:
            bc = tokens["border-color"]
            lines.append('  border: 0.75pt + rgb("{}"),'.format(bc))
            lines.append('  axis-stroke: 0.5pt + rgb("{}"),'.format(bc))
            lines.append('  grid-stroke: 0.3pt + rgb("{}"),'.format(bc))

        lines.append("  show-grid: true,")

        if "border-radius" in tokens:
            r = tokens["border-radius"]
            lines.append(f"  border-radius: {r}pt,")

        lines.append(")")
        lines.append("")

    return "\n".join(lines)


def generate_json(light, dark):
    """Generate JSON token file."""
    # Convert None to JSON null
    def clean(tokens):
        return {k: v for k, v in tokens.items()}

    return json.dumps({"light": clean(light), "dark": clean(dark)}, indent=2) + "\n"


# ── Main ──────────────────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(
        description="Extract CSS design tokens and generate primaviz theme files.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=__doc__.split("Usage:")[0],
    )
    parser.add_argument("css_file", help="Path to CSS file with custom properties")
    parser.add_argument("--out-dir", default="./typst", help="Output directory (default: ./typst)")
    parser.add_argument("--format", choices=["typst", "json", "both"], default="both",
                        help="Output format (default: both)")
    parser.add_argument("--name", default="theme", help="Theme name prefix (default: theme)")
    parser.add_argument("--dark-selector", default=".dark",
                        help="CSS selector for dark mode (default: .dark)")

    args = parser.parse_args()

    if not os.path.isfile(args.css_file):
        print(f"Error: file not found: {args.css_file}", file=sys.stderr)
        sys.exit(1)

    css = open(args.css_file).read()

    # Extract light (root) and dark tokens
    light_props = extract_vars(css, ":root")
    dark_props = extract_vars(css, args.dark_selector)

    if not light_props:
        print("Warning: no custom properties found in :root", file=sys.stderr)

    # Determine backgrounds for alpha blending
    light_bg_hex = "#FFFFFF"
    if "background" in light_props:
        bg = color_to_hex(light_props["background"])
        if bg:
            light_bg_hex = bg

    dark_bg_hex = "#0A0A0A"
    if "background" in dark_props:
        bg = color_to_hex(dark_props["background"])
        if bg:
            dark_bg_hex = bg

    light_tokens = build_tokens(light_props, light_bg_hex)
    dark_tokens = build_tokens(dark_props, dark_bg_hex)

    # Inherit light values for keys missing in dark (e.g., radius)
    for key in ("border-radius",):
        if key not in dark_tokens and key in light_tokens:
            dark_tokens[key] = light_tokens[key]

    print(f"Found {len(light_props)} light + {len(dark_props)} dark properties")
    print(f"Palette: {len(light_tokens.get('palette', []))} light, {len(dark_tokens.get('palette', []))} dark colors")

    # Create output directory
    os.makedirs(args.out_dir, exist_ok=True)

    if args.format in ("typst", "both"):
        path = os.path.join(args.out_dir, f"{args.name}.typ")
        with open(path, "w") as f:
            f.write(generate_typst(args.name, light_tokens, dark_tokens))
        print(f"Wrote {path}")

    if args.format in ("json", "both"):
        path = os.path.join(args.out_dir, f"{args.name}.json")
        with open(path, "w") as f:
            f.write(generate_json(light_tokens, dark_tokens))
        print(f"Wrote {path}")


if __name__ == "__main__":
    main()
